From c6acd484cae46b6f7c8cf267161ba1dc603b7816 Mon Sep 17 00:00:00 2001
From: Ilias Apalodimas <apalos@gmail.com>
Date: Tue, 21 Dec 2021 13:48:34 +0200
Subject: [PATCH] efi_loader: Don't authenticate certificates with CAs

This needs upstreaming and further discussion, but it seems like
vendors treat the MS cert as a self signed one ...

Signed-off-by: Ilias Apalodimas <apalos@gmail.com>
---
 lib/efi_loader/efi_signature.c | 18 ++++++++++++++----
 1 file changed, 14 insertions(+), 4 deletions(-)

diff --git a/lib/efi_loader/efi_signature.c b/lib/efi_loader/efi_signature.c
index 6e3ee3c0c004..e1d285f63838 100644
--- a/lib/efi_loader/efi_signature.c
+++ b/lib/efi_loader/efi_signature.c
@@ -480,11 +480,20 @@ bool efi_signature_verify(struct efi_image_regions *regs,
 			goto out;
 
 		EFI_PRINT("Verifying last certificate in chain\n");
-		if (signer->self_signed) {
-			if (efi_lookup_certificate(signer, db))
-				if (efi_signature_check_revocation(sinfo,
-								   signer, dbx))
+		/*
+		 * This needs disucssion in u-boot's ML and upstreaming
+		 * It seems that in practice no BIOS vendor chains certs
+		 * and expects the
+		 * 'Microsoft Corporation UEFI CA 2011' to be signed with
+		 * 'Microsoft Corporation Third Party Marketplace Root'
+		 *
+		 * This might fit into a Kconfig option
+		 */
+		/* if (signer->self_signed) { */
+		if (efi_lookup_certificate(signer, db))
+			if (efi_signature_check_revocation(sinfo, signer, dbx))
 					break;
+		/*
 		} else if (efi_verify_certificate(signer, db, &root)) {
 			bool check;
 
@@ -494,6 +503,7 @@ bool efi_signature_verify(struct efi_image_regions *regs,
 			if (check)
 				break;
 		}
+		*/
 
 		EFI_PRINT("Certificate chain didn't reach trusted CA\n");
 	}
-- 
2.30.2

